### ISSUE TITLE
Gen returning values outside of range

### ISSUE THREAD

**Issue by zbeckman** [typelift/SwiftCheck#89]
I have a short test that generates email addresses, and then feeds them to our UI. The UI then validates them against certain rules.

The problem is, the generators seem to be generating invalid data sets (meaning, Strings that they should *not* be generating). The rules do not include, for example, spaces and control characters. They also specify that the email should always contain a name, hostname, @, and a TLD.

But the tests generate data that fails on the following types of values:
```
09:54PM [Info] [default] [main] [TestEmailValidationRules.swift:147] spec() > generated '7_B.dt'
09:54PM [Info] [default] [main] [TestEmailValidationRules.swift:147] spec() > generated '_S@B.dt'
09:54PM [Info] [default] [main] [TestEmailValidationRules.swift:147] spec() > generated '7S@B.dt'
09:54PM [Info] [default] [main] [TestEmailValidationRules.swift:147] spec() > generated '.dA'
09:54PM [Info] [default] [main] [TestEmailValidationRules.swift:147] spec() > generated '.d
'
```
Here's the test case:

```swift
func wrap3(l : String) -> String -> String -> String -> String -> String {
	return { m in { m2 in { m3 in { r in l + m + m2 + m3 + r } } } }
}
			
fit("handle randomly generated addresses with Unicode") {
	let upperCase : Gen<Character>= Gen<Character>.fromElementsIn("A"..."Z")
	let lowerCase : Gen<Character> = Gen<Character>.fromElementsIn("a"..."z")
	let numeric : Gen<Character> = Gen<Character>.fromElementsIn("0"..."9")
	let alpha: Gen<Character> = Gen<Character>.oneOf([upperCase, lowerCase])
	let alphanumeric: Gen<Character> = Gen<Character>.oneOf([alpha, numeric])
	let latin: Gen<Character> = Gen<Character>.fromElementsIn("\u{00C0}"..."\u{02B8}")
	let allowedSymbols = Gen<Character>.fromElementsOf(["%", "+", "-", "_", "."])
	let special = Gen<Character>.fromElementsOf(["!", "#", "$", "%", "&", "'", "*", "+", "-", "/", "=", "?", "^", "_", "`", "{", "|", "}", "~", "."])

	let localEmail = Gen<Character>.oneOf([
		alphanumeric,
		allowedSymbols
		]).proliferateNonEmpty().suchThat({ $0.count > 2 && $0.count < 25 }).fmap(String.init)

	let hostname = Gen<Character>.oneOf([
		alpha
		]).proliferateNonEmpty().suchThat({ $0.count < 10 }).fmap(String.init)

	let tld = lower.proliferateNonEmpty().suchThat({ $0.count > 1 && $0.count < 4 }).fmap(String.init)

	let emailGen = wrap3 <^> localEmail <*> Gen.pure("@") <*> hostname <*> Gen.pure(".") <*> tld

	property("Unicode email addresses validate correctly") <- forAll(emailGen) { (e : String) in
		logger.info("generated '\(e)'")

	        f.text = e
		return f.isValid
	}
}
```

**Comment by CodaFi** [comment]
Since you seem to be having trouble with pods, I'd say check the git cache there and clear derived data.  Although this is probably not the best thing to say about a library based on generating random data: I cannot reproduce this.  The rules you've given should *never* give non-empty strings of non-printable characters.

**Comment by CodaFi** [comment]
Although I will note that the 2nd and 3rd examples you gave are actually valid email addresses.

**Comment by zbeckman** [comment]
I am using the *swift-develop* branch, since we are on Xcode 7 beta 6 (and Swift 2.0). I'm not using the pods since there is no pod config in this branch (yet). I've cleared derived data and cleaned my project directory half a dozen times now...

My Cartfile:
```
# SwiftCheck
git "https://github.com/typelift/SwiftCheck.git" "swift-develop"
```
And yes, 2nd and 3rd are good... I just grabbed a small sample. I'm also getting a few cases that just make absolutely no sense whatsoever:
```
12:26PM [Info] [default] [main] [TestEmailValidationRules.swift:147] spec() > generated '73B

'
12:26PM [Info] [default] [main] [TestEmailValidationRules.swift:147] spec() > generated '2_2



'
12:26PM [Info] [default] [main] [TestEmailValidationRules.swift:147] spec() > generated '



'
12:26PM [Info] [default] [main] [TestEmailValidationRules.swift:147] spec() > generated '



'
12:26PM [Info] [default] [main] [TestEmailValidationRules.swift:147] spec() > generated '1




'
```

**Comment by CodaFi** [comment]
What happens when you ask each component generator to `.generate` a value for you?

**Comment by zbeckman** [comment]
Aha! Working on a simpler repro for you, and I found one... by accident.

I put together a simple random uppercase string generator. By mistake, I put in a regex match for "not A-Z" instead of "A-Z" which, of course, results in everything failing.

But guess what...? SwiftCheck diligently tried to reduce the problem and find out what was failing so it started generating a *huge* number of test cases. If you just generate 100, you won't see this unless you are very, very lucky. But after 1,000 it crops up pretty frequently.

So here's the test case – this will never succeed, and SwiftCheck will try and try again forever...
```
fit("generates simple random sets") {
	let stringGen: Gen<String> = Gen<Int>.fromElementsIn(10...20) >>- { size in
		return Gen<Character>.fromElementsIn("A"..."Z").proliferateSized(size).fmap(String.init)
	}
				
	property("generated strings are within spec") <- forAll(stringGen) { (e : String) in
		logger.info("string -> '\(e)'")
		return (e =~ "[^A-Z]")
	}
}
```
And take a look at the output:
```
03:51PM [Info] [testing] [main] [TestGenerators.swift:87] spec() > string -> 'ARREPQCAAAAAAAAAAA'
03:51PM [Info] [testing] [main] [TestGenerators.swift:87] spec() > string -> '
RREPQCAAAAAAAAAAA'
03:51PM [Info] [testing] [main] [TestGenerators.swift:87] spec() > string -> '3RREPQCAAAAAAAAAAA'
03:51PM [Info] [testing] [main] [TestGenerators.swift:87] spec() > string -> '1RREPQCAAAAAAAAAAA'
03:51PM [Info] [testing] [main] [TestGenerators.swift:87] spec() > string -> 'CRREPQCAAAAAAAAAAA'
03:51PM [Info] [testing] [main] [TestGenerators.swift:87] spec() > string -> '2RREPQCAAAAAAAAAAA'
03:51PM [Info] [testing] [main] [TestGenerators.swift:87] spec() > string -> 'BRREPQCAAAAAAAAAAA'
03:51PM [Info] [testing] [main] [TestGenerators.swift:87] spec() > string -> 'X REPQCAAAAAAAAAAA'
03:51PM [Info] [testing] [main] [TestGenerators.swift:87] spec() > string -> 'XAREPQCAAAAAAAAAAA'
03:51PM [Info] [testing] [main] [TestGenerators.swift:87] spec() > string -> 'X
REPQCAAAAAAAAAAA'
03:51PM [Info] [testing] [main] [TestGenerators.swift:87] spec() > string -> 'X1REPQCAAAAAAAAAAA'
```
The `=~` operator is a simple regex matcher I put together. Here's a gist for it: https://gist.github.com/zbeckman/ef7ea36d70186e4dfdc5 but it doesn't really matter, just use `return false` and you'll end up in the same place.

And just to be clear: If you use `return (e =~ "[A-Z]")` you will *probably never see the problem* unless you are very lucky. I suspect it has something to do with SwiftCheck trying to reduce the problem... somewhere in there, it must be injecting a few values that it shouldn't. Trying too hard...? ;-)

Oh by the way: There seems to be a recurring pattern. With the output above if you start scrolling down the output, you'll notice that a space gets injected pretty regularly, and the same with the newline...

**Comment by CodaFi** [comment]
Because the shrinker for Strings gets invoked.  You have no control over that!

**Comment by CodaFi** [comment]
Try using the `.noShrinking` combinator if this comes up again.

**Comment by zbeckman** [comment]
Ok. This does seem to be the same thing that happens in the original case though – probably because a significant number of patterns that get generated were rejected by our email validator (which actually tests against W3C email rules). So – some rejects probably led to some shrinkage. But shouldn't the shrinker not be injecting spaces and newlines? I'm puzzled by that.

**Comment by CodaFi** [comment]
The `Arbitrary` constraint on `forAll` is so you don't have to provide a shrinker, one is always given for you by default and you have no control over what it is.  IIRC `scalacheck` fixes this by providing a set of `forAllNoShrink` combinators, which is probably what we should do too (but I'm wary because of the property modifier I just listed above).  If we just used the generator, shrinking will always fail to terminate because it will continue to expand on to infinity.

**Comment by CodaFi** [comment]
Basically the shrinker doesn't know about your generators and doesn't care.  It only cares about finding a minimal failing test case.

**Comment by zbeckman** [comment]
Hmmm. I see... explains why it doesn't happen in Scala. Well, a little more info – and it does pose a nasty problem. Updated code:
```swift
fit("generates simple random sets") {
	let stringGen: Gen<String> = Gen<Int>.fromElementsIn(10...20) >>- { size in
		return Gen<Character>.fromElementsIn("A"..."Z").proliferateSized(size).fmap(String.init)
	}
				
	property("generated strings are within spec") <- forAll(stringGen) { (e : String) in
		logger.info("string -> '\(e)'")
		return ((e =~ "[A-Z]") && e.characters.count >= 10 && e.characters.count <= 20)
	}.noShrinking
}
```
But check the output, within the first 50 runs:
```
04:06PM [Info] [testing] [main] [TestGenerators.swift:87] spec() > string -> 'ARRLEPQCXQIHJFAFLOR'
04:06PM [Info] [testing] [main] [TestGenerators.swift:87] spec() > string -> '
RRLEPQCXQIHJFAFLOR'
04:06PM [Info] [testing] [main] [TestGenerators.swift:87] spec() > string -> '3RRLEPQCXQIHJFAFLOR'
```
Even with no shrinking, we got a newline thrown in there. Bleah. That poses a problem since what we need is a reliable 10-20 character generate... no spaces...

(Sorry if I'm being obtuse on the root cause here – still pretty new to SwiftCheck and still trying to figure out it's capabilities).

**Comment by CodaFi** [comment]
It doesn't happen because, despite both names being the same, the semantics of the two frameworks are *very* different.  ScalaCheck has made certain compromises that enable, IMO, very productive testing at the expense of simplicity and ease of use.

**Comment by zbeckman** [comment]
Agree on how hard it was to cut my teeth on ScalaCheck. ;-)

However – I'm a bit stuck at this point. I don't see how I can generate our test data... that is, based on the generator above, it would appear that those random newline/space characters are being injected (even when I use `noShrinking`, and even though it is pretty rare). And that poses a problem because of such characters are injected it breaks many of our test conditions...

**Comment by CodaFi** [comment]
One would argue that it is your responsibility to fix your tests and not ours to fix the shrinker ;)

**Comment by CodaFi** [comment]
.  If all else fails write a wrapper type for string, make it conform to arbitrary (see the modifier types for an example) and write your own shrinker.

**Comment by zbeckman** [comment]
Well... maybe. The point of a testing framework is to find and help diagnose problems... so yes, of course we want good test cases – but isn't part of the problem identifying and reducing the problem, so that we can identify those cases? (I don't want to get into a philosophical argument over it... I'm just saying that it's *relatively harder* to diagnose the problem with this behavior in the system).

So yes, I can fix the test cases as and when we uncover problems. But it exposes a risk because when I use `forAll` I won't know if my generators are whacky, or my program is failing. ;-)

By the way, earlier you asked what happens if I just use `generate`, and the results are as expected. So clearly the issue is shrinker related. The following works fine:
```swift
fit("generates simple random sets") {
	let stringGen: Gen<String> = Gen<Int>.fromElementsIn(10...20) >>- { size in
		return Gen<Character>.fromElementsIn("A"..."Z").proliferateSized(size).fmap(String.init)
	}
				
	for _ in 0...1000 {
		let e = stringGen.generate
		logger.info("string -> '\(e)'")
		expect((e =~ "[A-Z]") && e.characters.count >= 10 && e.characters.count <= 20).to(beTrue())
	}
```
But not I'm not benefitting from the shrinker or `forAll` and much of SwiftCheck's power.

Coming from ScalaCheck... Hm, I think I would say I appreciate their approach to shrinking.

Anyhow – This would be an *excellent* topic to put into the documentation. Now that I understand it, I can work around it (write my own shrinker, perhaps, as you say). But it was a very unexpected behavior. I'm surprised others haven't stumbled on it and reported it as a bug as well.

Thanks for all the quick turnaround on explaining it!

**Comment by zbeckman** [comment]
Oh... But... I am wondering why using `.noShrinking` didn't actually work...

**Comment by CodaFi** [comment]
Sorry to be so glib, but this behaves as intended (we take QuickCheck's semantics here).

However that is an interesting question.  Where are you applying it?  To the inside test case or the outside forAll?

**Comment by zbeckman** [comment]
See the example up above (8 posts back).

**Comment by CodaFi** [comment]
Hm... It should quantify over everything inside... Maybe try inside explicitly?

**Comment by zbeckman** [comment]
Err... Not understanding that syntax:
```swift
return noShrinking((e =~ "[A-Z]") && e.characters.count >= 10 && e.characters.count <= 20)
```
Is clearly not working.

**Comment by CodaFi** [comment]
Surround the thing in parens and use it as a property getter.

**Comment by zbeckman** [comment]
No good:
```swift
property("generated strings are within spec") <- (forAll(stringGen) { (e : String) in
	logger.info("string -> '\(e)'")
	return ((e =~ "[A-Z]") && e.characters.count >= 10 && e.characters.count <= 20)
}).noShrinking
```
Results still have shrinking artifacts:
```
05:01PM [Info] [testing] [main] [TestGenerators.swift:93] spec() > string -> 'LVWHOXXVA EFKLQEQ'
05:01PM [Info] [testing] [main] [TestGenerators.swift:93] spec() > string -> 'LVWHOXXVAAEFKLQEQ'
05:01PM [Info] [testing] [main] [TestGenerators.swift:93] spec() > string -> 'LVWHOXXVA
EFKLQEQ'
05:01PM [Info] [testing] [main] [TestGenerators.swift:93] spec() > string -> 'LVWHOXXVA3EFKLQEQ'
```

**Comment by CodaFi** [comment]
Hmph... It doesn't make any sense.  No Shrink literally replaces any shrinking results with an empty array...

**Comment by CodaFi** [comment]
:exclamation: I think I know what's happening here! :exclamation: 

OK, so Haskell's QuickCheck, in its infinite laziness, has a point where it recursively requests the properties of failing properties of failing properties... etc. until it reaches and ending case.  From there it holds a tree of thunks that get forced one-by-one as their shrinking loop happens to get to them.  But because Swift is eager it automatically forces the entirety of the array of generated thunk values which, in turn, invoke the property test with "spurious" information.  The good part about all of this is that those bogus test cases aren't actually affecting your final outcome, you're just failing on the case SwiftCheck is showing you and everything is working, even `noShrinking`.

tl;dr The shrinker loop runs regardless of whether you asked it to or not, but when you ask it not to it doesn't penalize you for failing on shrunk cases, it just discards the result.

So yeah, it's not a bug, but it's definitely something to look at for future optimizations.

I recommend, if you ever need to log anything from your properties, that you look into `withCallback`, `whenFail` and `whenEachFail`.  Each of those *only* gets called at the end after SwiftCheck has finished evaluating all the branches and come to a proper result.  Print from in there (especially from `withCallback`) and you'll get *real* data.

**Comment by CodaFi** [comment]
I'm going to close this now.  I'll make an effort to include this as part of the documentation update.
